src/offset/local/unix.rs

```diff
@@ -31,69 +31,70 @@ thread_local! {
+    if let Ok(s) = env_result {
+        // this avoids bringing in a dependency for calculating the hash
+        let mut hasher = hash_map::DefaultHasher::new();
+        hasher.write(s.as_bytes());
+        Some(hasher.finish())
+    } else {
+        None
+    }
+}
+
 enum Source {
-    LocalTime { mtime: SystemTime, last_checked: SystemTime },
-    // we don't bother storing the contents of the environment variable in this case.
-    // changing the environment while the process is running is generally not reccomended
-    Environment,
+    LocalTime { mtime: SystemTime },
+    Environment { hash: u64 },
 }
 
-impl Default for Source {
-    fn default() -> Source {
-        // use of var_os avoids allocating, which is nice
-        // as we are only going to discard the string anyway
-        // but we must ensure the contents are valid unicode
-        // otherwise the behaivour here would be different
-        // to that in `naive_to_local`
-        match env::var_os("TZ") {
-            Some(ref s) if s.to_str().is_some() => Source::Environment,
-            Some(_) | None => match fs::symlink_metadata("/etc/localtime") {
+impl Source {
+    fn new(env_result: &Result<String, env::VarError>) -> Source {
+        match get_tz_env_hash(env_result) {
+            Some(hash) => Source::Environment { hash },
+            None => match fs::symlink_metadata("/etc/localtime") {
                 Ok(data) => Source::LocalTime {
                     // we have to pick a sensible default when the mtime fails
                     // by picking SystemTime::now() we raise the probability of
                     // the cache being invalidated if/when the mtime starts working
                     mtime: data.modified().unwrap_or_else(|_| SystemTime::now()),
-                    last_checked: SystemTime::now(),
                 },
                 Err(_) => {
                     // as above, now() should be a better default than some constant
                     // TODO: see if we can improve caching in the case where the fallback is a valid timezone
-                    Source::LocalTime { mtime: SystemTime::now(), last_checked: SystemTime::now() }
+                    Source::LocalTime { mtime: SystemTime::now() }
                 }
             },
         }
     }
-}
-
-impl Source {
-    fn out_of_date(&mut self) -> bool {
-        let now = SystemTime::now();
-        let prev = match self {
-            Source::LocalTime { mtime, last_checked } => match now.duration_since(*last_checked) {
-                Ok(d) if d.as_secs() < 1 => return false,
-                Ok(_) | Err(_) => *mtime,
-            },
-            Source::Environment => return false,
-        };
-
-        match Source::default() {
-            Source::LocalTime { mtime, .. } => {
-                *self = Source::LocalTime { mtime, last_checked: now };
-                prev != mtime
+    fn out_of_date(&self, other_source: &Source) -> bool {
```

Comment:
Add an empty line before this? I'd also rename `other_source` as `prev` or something, since you're already using names like `old_mtime` and `old_hash` for some of the inner data it would make sense to use the more specific semantics here.

Reply:
I've ended up in-lining this as above, which resulted in a nicer progression from the old code to the new code